tools/stats/monitor.py

```diff
@@ -76,78 +59,44 @@ def get_per_process_gpu_info(handle: Any) -> List[Dict[str, Any]]:
     return per_process_info
 
 
-def rocm_ret_ok(ret: int) -> Any:
-    return ret == rsmi_status_t.RSMI_STATUS_SUCCESS
-
-
-def rocm_list_devices() -> List[int]:
-    num = c_uint32(0)
-    ret = rocmsmi.rsmi_num_monitor_devices(byref(num))
-    if rocm_ret_ok(ret):
-        return list(range(num.value))
-    return []
-
-
-def rocm_get_mem_use(device: int) -> float:
-    memoryUse = c_uint64()
-    memoryTot = c_uint64()
-
-    ret = rocmsmi.rsmi_dev_memory_usage_get(device, 0, byref(memoryUse))
-    if rocm_ret_ok(ret):
-        ret = rocmsmi.rsmi_dev_memory_total_get(device, 0, byref(memoryTot))
-        if rocm_ret_ok(ret):
-            return float(memoryUse.value) / float(memoryTot.value)
-    return 0.0
-
-
-def rocm_get_gpu_use(device: int) -> float:
-    percent = c_uint32()
-    ret = rocmsmi.rsmi_dev_busy_percent_get(device, byref(percent))
-    if rocm_ret_ok(ret):
-        return float(percent.value)
-    return 0.0
-
-
-def rocm_get_pid_list() -> List[Any]:
-    num_items = c_uint32()
-    ret = rocmsmi.rsmi_compute_process_info_get(None, byref(num_items))
-    if rocm_ret_ok(ret):
-        buff_sz = num_items.value + 10
-        procs = (rsmi_process_info_t * buff_sz)()
-        procList = []
-        ret = rocmsmi.rsmi_compute_process_info_get(byref(procs), byref(num_items))
-        for i in range(num_items.value):
-            procList.append(procs[i].process_id)
-        return procList
-    return []
-
-
-def rocm_get_per_process_gpu_info() -> List[Dict[str, Any]]:
+def rocm_get_per_process_gpu_info(handle: Any) -> List[Dict[str, Any]]:
+    processes = pyamdsmi.amdsmi_get_gpu_process_list(handle)
     per_process_info = []
-    for pid in rocm_get_pid_list():
-        proc = rsmi_process_info_t()
-        ret = rocmsmi.rsmi_compute_process_info_by_pid_get(int(pid), byref(proc))
-        if rocm_ret_ok(ret):
-            info = {"pid": pid, "gpu_memory": proc.vram_usage}
-            per_process_info.append(info)
+    for p in processes:
+        proc_info = pyamdsmi.amdsmi_get_gpu_process_info(handle, p)
+        info = {
+            "pid": proc_info["pid"],
+            "gpu_memory": proc_info["memory_usage"]["vram_mem"],
+        }
+        per_process_info.append(info)
     return per_process_info
 
 
 if __name__ == "__main__":
     handle = None
     try:
-        pynvml.nvmlInit()
-        handle = pynvml.nvmlDeviceGetHandleByIndex(0)
-    except pynvml.NVMLError:
+        import pynvml  # type: ignore[import]
+
+        try:
+            pynvml.nvmlInit()
+            handle = pynvml.nvmlDeviceGetHandleByIndex(0)
+        except pynvml.NVMLError:
+            pass
+    except ModuleNotFoundError:
```

Comment:
I think it's better to avoid the nested try/except here.  The result is the same since both except clauses just pass.

```suggestion
        import pynvml  # type: ignore[import]
        pynvml.nvmlInit()
        handle = pynvml.nvmlDeviceGetHandleByIndex(0)
    except (ModuleNotFoundError, pynvml.NVMLError):
        pass
```

Reply:
Ah so now I remember why I went this route initially. If we go this way we would need both amdsmi and pynvml installed in both AMD/Nvidia environment as it can import pynvml.NVMLError without pynvml installed

Traceback (most recent call last):
  File "/root/pytorch/tools/stats/monitor.py", line 78, in <module>
    import pynvml  # type: ignore[import]
ModuleNotFoundError: No module named 'pynvml'

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "/root/pytorch/tools/stats/monitor.py", line 82, in <module>
    except (ModuleNotFoundError, pynvml.NVMLError):
NameError: name 'pynvml' is not defined
```

Same would work the other way around, we'd need amdsmi installed on NV otherwise we cannot add AmdSmiException to the except. AFAIK nested try blocks would be the only way around this. Unless it's acceptable to only use `except ModuleNotFoundError` for both cases here.